libs/domain/checkout/src/services/default-checkout-data.service.ts

```diff
@@ -1,155 +1,189 @@
+    return this.cartService.isEmpty().pipe(map((isEmpty) => isEmpty === false));
+  }
 
-    this.subscription = this.identity
-      .get()
+  /**
+   * Restores the data from storage.
+   *
+   * TODO: consider storage strategies
+   */
+  protected restore(): void {
+    this.storage
+      .get<Checkout>(checkoutDataStorageKey, StorageType.SESSION)
       .pipe(
-        skip(1),
-        // we want to reset checkout state every time user will change
-        tap(() => this.reset())
+        take(1),
+        tap((stored) => {
+          if (stored) {
+            const map = new Map<
+              keyof Checkout,
+              { data: Checkout[keyof Checkout] | null; valid: boolean }
+            >();
+            Object.keys(stored).forEach(() => {
+              const entries = Object.entries(stored);
+              for (const [key, value] of entries) {
+                map.set(key as keyof Checkout, {
+                  data: value as Checkout[keyof Checkout] | null,
+                  valid: false,
+                });
+              }
+            });
+            this.subject.next(map);
+          }
+        })
       )
       .subscribe();
   }
 
-  isGuestCheckout(): Observable<boolean> {
-    return this.isGuestCheckout$;
-  }
-
-  setGuestCheckout(isGuestCheckout = true): Observable<unknown> {
-    if (isGuestCheckout) {
-      this.isGuestCheckout$.next(true);
-      this.storage.set(guestCheckoutStorageKey, true, StorageType.SESSION);
-    } else {
-      this.reset();
-    }
-
-    return this.isGuestCheckout$;
-  }
-
-  reset(): void {
-    this.isGuestCheckout$.next(false);
-    this.addressDetails$.next(null);
-    this.contactDetails$.next(null);
-    this.shipmentDetails$.next(null);
-    this.paymentDetails$.next(null);
-    this.storage.remove(guestCheckoutStorageKey, StorageType.SESSION);
-    this.storage.remove(addressCheckoutStorageKey, StorageType.SESSION);
-    this.storage.remove(contactCheckoutStorageKey, StorageType.SESSION);
-    this.storage.remove(shipmentCheckoutStorageKey, StorageType.SESSION);
-    this.storage.remove(paymentCheckoutStorageKey, StorageType.SESSION);
-  }
-
-  getCustomer(): Observable<ContactDetails | null> {
-    return this.contactDetails$;
-  }
-
-  setCustomer(contactDetails: ContactDetails | null): Observable<void> {
-    this.contactDetails$.next(contactDetails);
-    return this.storage.set(
-      contactCheckoutStorageKey,
-      contactDetails,
-      StorageType.SESSION
+  protected cartId = this.cartService.getCart({}).pipe(
+    map((cart) => cart?.id),
+    tap((cartId) => {
+      const data = cartId ?? null;
+      this.set('cartId', !!data, data);
+    })
+  );
+
+  protected load$ = this.cartId.pipe(
+    switchMap((cartId) =>
+      cartId
+        ? this.adapter.get({
+            cartId,
+            // TODO: consider making this configurable by the type of checkout we're using
+            include: [
+              ApiCheckoutModel.Includes.Shipments,
```

Comment:
Ideally we will move backend specific details (and usage of backend models) to adapters